### PR TITLE
feat: code editor export

### PR DESCRIPTION
**Pull request by gratcliff**
## 🧰 What's being changed?

- Adding `react-codemirror2` as core dep.
- Refactoring directory structure for common usage
- Using `esm module syntax` throughout
- New Core Export via `options` -> Full `CodeMirror` Instance! Usage through `{ editable: true }`
- Codemirror `runmode` highlighting + style + line numbers via `{ highlightMode: true }`

## 🧪 Testing

Can run locally via `npm run start` to see our full `react-codemirror2` instance in action.

### CODE REVIEW

**Comment by gratcliff** [comment]
@rafegoldberg Would love your support here (or in a separate branch/PR) teasing out duplicate styles.

**Comment by gratcliff** [comment]
Just realized that I need to work on adding a `readme.md` to this! Working on it now!

File: __tests__/codeEditor.test.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    expect(node.props().lang).toBe('js');
+    expect(cm.props().options.mode).toBe('javascript');
+  });
+
+  it('should set new language via props', () => {
+    node.setProps({ lang: 'kotlin' });
+    expect(node.props().lang).toBe('kotlin');
+
+    setTimeout(() => {
+      console.log(cm.props());
```
Comment (erunion) [inline __tests__/codeEditor.test.js:35]:
```suggestion
```

File: src/index.js [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
+
+  return React.createElement(
+    'div',
+    { className: classes },
     codemirror(typeof code === 'string' ? code : '', lang, opts)
   );
+};
 
-module.exports.uppercase = require('./uppercase');
+module.exports.uppercase = require('./utils/uppercase');
```
Comment (erunion) [inline src/index.js:33]: Might as well add an export for `canonical` while we're at it.

```suggestion
module.exports.uppercase = require('./utils/uppercase');
module.exports.canonical = require('./utils/canonical');
```

File: src/utils/modes.js [hunk 3]
```diff
@@ -9,7 +9,31 @@
 // `htmlmixed`, but `java`, needs to be rendered using the `clike` mode.
 //
 // We also have the mimeType to potentially in future load in new types dynamically.
-module.exports = {
+require('codemirror/mode/clike/clike');
```
Comment (erunion) [inline src/utils/modes.js:12]: Can you update the location of `codemirror.jsx` in the comment on line 2?

File: src/index.js [hunk 4]
```diff
@@ -1,14 +1,33 @@
...
+        code,
+        lang,
+      },
+      null
+    );
+  }
+
+  let classes = '';
+  if (opts.highlightMode) {
+    classes = 'CodeEditor cm-s-material-palenight';
```
Comment (erunion) [inline src/index.js:21]: If `editable` is enabled, should these classes be applied to that element?

File: src/codeMirror/index.jsx [hunk 5]
```diff
@@ -0,0 +1,151 @@
+const CodeMirror = require('codemirror');
+const React = require('react');
+const PropTypes = require('prop-types');
+const Variable = require('@readme/variable');
+const { getMode } = require('../utils/modes');
```
Comment (erunion) [inline src/codeMirror/index.jsx:5]: I assume loading these modes before `codemirror/addon/runmode/runmode` works just fine and languages like PHP and whatnot are still recognized?

File: package.json [hunk 6]
```diff
@@ -6,7 +6,9 @@
   "dependencies": {
     "@readme/variable": "^7.2.1",
     "codemirror": "^5.48.2",
+    "prop-types": "^15.7.2",
     "react": "^16.13.1",
+    "react-codemirror2": "^7.2.1",
```
Comment (erunion) [inline package.json:11]: Where's this dep being used?

File: src/index.js (hunk 4 above)
Comment (gratcliff) [reply src/index.js:21]: The `editable` flag transitions the output to use the full `react-codemirror2` export, which has a bunch of styling baked in. We don't need to apply the same classes to both.

File: src/utils/modes.js (hunk 3 above)
Comment (gratcliff) [reply src/utils/modes.js:12]: Done!

File: src/index.js (hunk 2 above)
Comment (gratcliff) [reply src/index.js:33]: Good call, done!

File: __tests__/codeEditor.test.js (hunk 1 above)
Comment (gratcliff) [reply __tests__/codeEditor.test.js:35]: Good catch, removed.

File: src/codeMirror/index.jsx (hunk 5 above)
Comment (gratcliff) [reply src/codeMirror/index.jsx:5]: Indeed they do! Feel free to check it out when you have a sec!

File: package.json (hunk 6 above)
Comment (gratcliff) [reply package.json:11]: Right [here](https://github.com/readmeio/syntax-highlighter/blob/feat/codeeditor-export/src/codeEditor/index.jsx#L3)!

File: src/codeMirror/index.jsx (hunk 5 above)
Comment (erunion) [reply src/codeMirror/index.jsx:5]: Screen Shot 2020-09-14 at 2 02 23 PM

sweet

**Review by erunion — APPROVED** [review]
lgtm! feel free to merge when tests are all good. lmk and i'll tag a major version bump for you

**Comment by gratcliff** [comment]
> lgtm! feel free to merge when tests are all good. lmk and i'll tag a major version bump for you

Thanks a lot @erunion! I really appreciate the review. Let me get some more eyes on this before we go for a release!

**Comment by erunion** [comment]
When we go for a new release we should also remove this library from the api-explorer repo too since it's still there. I dinked around a bit with that two weeks ago but ran into a bunch of Jest compilation issues that are probably no longer a problem with the work here.

**Comment by gratcliff** [comment]
> When we go for a new release we should also remove this library from the api-explorer repo too since it's still there. I dinked around a bit with that two weeks ago but ran into a bunch of Jest compilation issues that are probably no longer a problem with the work here.

Sounds perfect! Next steps for this are integrating in the `api-explorer` and the `ui` libs.

File: __tests__/codeMirror.test.js [hunk 7]
```diff
@@ -14,7 +11,7 @@ test('should highlight a block of code', () => {
 
   expect(code.hasClass('cm-s-neo')).toBe(true);
   expect(code.html()).toBe(
-    '<span class="cm-s-neo"><span class="cm-keyword">var</span> <span class="cm-def">a</span> <span class="cm-operator">=</span> <span class="cm-number">1</span>;</span>'
+    '<div class="cm-s-neo"><span class="cm-keyword">var</span> <span class="cm-def">a</span> <span class="cm-operator">=</span> <span class="cm-number">1</span>;</div>'
```
Comment (rafegoldberg) [inline __tests__/codeMirror.test.js:14]: We might want to refactor these HTML expectations in to snapshots at some point.

File: src/codeEditor/index.jsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+CodeEditor.propTypes = {
+  children: PropTypes.string,
+  className: PropTypes.string,
+  /** Code to show in the editor.
+   */
+  code: PropTypes.string,
+  /** CodeMirror language mode to use.
+   */
+  lang: PropTypes.string,
+  /** Custom CodeMirror  configuration.
```
Comment (rafegoldberg) [inline src/codeEditor/index.jsx:45]:
```suggestion
  /** Custom CodeMirror configuration.
```

File: src/codeMirror/index.jsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  CodeMirror.runMode(code, mode, (text, style) => {
+    if (style !== curStyle) {
+      flush();
+      curStyle = style;
+      accum = text;
+    } else {
+      accum += text;
+    }
+  });
+  flush();
```
Comment (rafegoldberg) [inline src/codeMirror/index.jsx:120]: There are three different potential return values, right? Might be nice to break this up in to multiple methods, or add some comments above each section just so we don't forget down the line!

File: src/codeMirror/style.scss [hunk 10]
```diff
@@ -0,0 +1,53 @@
+$gutterWidth: 43px; // enough for 3 chars
```
Comment (rafegoldberg) [inline src/codeMirror/style.scss:1]: ~Not sure if this a Tony thang or not, but we should probably be using `em` values if we're trying to approximate font widths.~ Never mind — I see we're doing further SCSS-based calculations on this value, so all the math would have to be converted to CSS `calc()`, and at that point it's probably outside the scope of this PR!

File: webpack.config.js [hunk 11]
```diff
@@ -14,11 +15,36 @@ module.exports = {
...
+        use: ['style-loader', 'css-loader', 'sass-loader'],
+      },
     ],
   },
   optimization: {
     minimize: true,
-    minimizer: [new TerserPlugin()],
+    minimizer: [
+      new TerserPlugin({
+        sourceMap: true,
```
Comment (erunion) [inline webpack.config.js:28]: Should this be enabled for the production config?

**Review by erunion — APPROVED** [review]

File: webpack.config.js (hunk 11 above)
Comment (gratcliff) [reply webpack.config.js:28]: Fair comment! I'll rip it out, it's not helping anyway.

File: src/index.js [hunk 12]
```diff
@@ -1,14 +1,36 @@
...
+      lang,
+    });
+  }
+
+  let classes = '';
+  if (opts.highlightMode) {
+    classes = 'CodeEditor cm-s-material-palenight';
+  } else {
+    classes = opts.dark ? 'cm-s-tomorrow-night' : 'cm-s-neo';
+  }
```
Comment (rafegoldberg) [inline src/index.js:26]: I think we should make a decision between these two dark themes regardless of rendering strategy. (I would personally vote for Material Palenight!) Thoughts?

File: src/utils/modes.js [hunk 13]
```diff
@@ -1,5 +1,5 @@
 // This is a mapping of languages that we support, but aren't directly loading the mode extension
-// for within `codemirror.jsx`.
+// for within `/src/codeMirror/index.jsx` and /src/codeEditor.index.jsx
```
Comment (rafegoldberg) [inline src/utils/modes.js:2]:
```suggestion
// within `/src/codeMirror/index.jsx` and /src/codeEditor.index.jsx
```

**Review by rafegoldberg — APPROVED** [review]
Just some minor clean up stuff and a couple of questions. Otherwise I'm in to this!

File: __tests__/codeMirror.test.js (hunk 7 above)
Comment (gratcliff) [reply __tests__/codeMirror.test.js:14]: Agreed, or vice versa!

File: src/index.js (hunk 12 above)
Comment (gratcliff) [reply src/index.js:26]: We should, I think this should be bundled into a large decision/work to consolidate the css between this, the `ui`, and `readme` libs.

File: src/codeMirror/index.jsx (hunk 9 above)
Comment (gratcliff) [reply src/codeMirror/index.jsx:154]: We should only have 2 variant return values:
- legacy
- More traditional JSX/ DOM structure (`divs`, styling, etc). 

They both rely on the same core to a large extent - but I broke out the styled part into it's own functional component!

**Pull request merged**